Skip invalid data instead of exiting completely.

basememara opened the pull request:
Whether using `Unbox` or `UnboxOrThrow`, the current implementation nil's the object or returns an empty collection completely if there is a single property on a single object that is invalid. This is necessary when perfect data is needed.

However, this pull request adds an option to skip invalid data and move on to the next property or object. This way, the consuming developer can prefer to deal with default values instead of an empty collection or nil object.

Use as:
```
let items: [Post] = UnboxOrSkip(data)
```

Also part of discussion #39.

JohnSundell commented on Unbox.swift:
```diff
@@ -313,14 +371,17 @@ public class Unboxer {
     /// Any contextual object that was supplied when unboxing was started
     public let context: Any?
+    /// Allow Unboxer to continue on invalid data
+    public let skipInvalidData: Bool
```
Does this information need to be exposed to the API user?

basememara replied:
No I wouldn't think so, it can be private which is better.

JohnSundell commented on Unbox.swift:
```diff
@@ -649,26 +710,26 @@ private extension UnboxableWithContext {
 
 private extension Unboxer {
-    static func unboxerFromData(data: NSData, context: Any?) throws -> Unboxer {
+    static func unboxerFromData(data: NSData, context: Any?, skipInvalidData: Bool = false) throws -> Unboxer {
```
Better to not use default values in the implementation, since it forces you to make a decision whether to include it or not (since the convenience aspect of it isn't really relevant for implementation details, only for public APIs).

basememara replied:
I see, that makes sense and seems more conscious. I just didn't want to impact the rest of the code.

basememara commented on Unbox.swift:
```diff
@@ -49,7 +49,7 @@ public func Unbox<T: Unboxable>(data: NSData, context: Any? = nil) -> T? {
 
 /// Unbox binary data into an array of `T`, while optionally using a contextual object
-public func Unbox<T: Unboxable>(data: NSData, context: Any? = nil) -> [T]? {
+public func Unbox<T: Unboxable>(data: NSData, context: Any? = nil, skipInvalidData: Bool = false) -> [T]? {
```
Good catch, that was a mistake left from previous attempt before going with `UnboxOrSkip` instead of overloading other functions.

JohnSundell replied:
I actually think this is the better approach :) See my comment on the PR itself.

JohnSundell commented:
Thanks for this PR! :rocket:  I think it's a great addition. :+1:

However, I think it'd be a lot better if you continued the parameter approach (extending `Unbox() -> [T]` rather than introducing a new `UnboxOrSkip()` function. Adding a lot of new `Unbox...` main functions will make the API more complicated and harder to maintain, since every time something is changed all the functions have to be updated. And since it's only relevant for arrays anyway, it makes sense to implement it as a parameter.

Also, `skipInvalidData` kind of sounds like you will skip on an entire invalid array, rather than elements of the array. The intent here is to surpress errors in case an error was encountered while unboxing a member of the array, so how about naming it `allowInvalidElements: Bool = true` (this matches the `Element` naming convenion of `Array`. What do you think?

basememara commented:
Yes I actually toyed with both ways, as an overloaded parameter and using a separate `UnboxOrSkip`. In retrospect, using an overloaded parameter on `Unbox` seems like the way to go too because it is indeed a parameter and overkill to create a new function. It doesn't change the overall implementation or return types (returning a `nil` list is better than empty one, it gives more of an indication of what happened).

I never liked `skipInvalidData` name. I really really like `allowInvalidElements`.

Let me make the changes and add some unit tests and I'll update the PR. Thanks a mil for the input!

JohnSundell commented:
Awesome! Thank you so much for doing this :+1:

basememara commented:
I just updated the PR to include the updates we discussed and it feels better. The only thing I don't like is I was forced to add the `allowInvalidElements` parameter to `UnboxOrThrow` as well so it can be reused for `Unbox`. I think that defeats the purpose of throwing errors since now what possible errors can `UnboxOrThrow` throw? In other words, does this even make sense?: `UnboxOrThrow(data, allowInvalidElements: true)`

I've also added a unit test. I couldn't figure how to reuse your mock data so I just generated dummy JSON and unboxed it.

Let me know if you have any thoughts or suggestions from here and I'll be glad to do another iteration on it. Thanks again!

basememara commented:
Ok not sure how you're going to feel about this, but I moved the public API logic into private functions of `Unboxer`. This way, the public API functions are purely wrappers for them and refactoring can happen in the private functions instead of the public `UnboxOrThrow` functions. This also resolved `UnboxOrThrow` having the `allowInvalidElements` parameter.

JohnSundell commented on Unbox.swift:
```diff
@@ -696,12 +691,64 @@ private extension Unboxer {
     }
+
+    /// Unbox a JSON dictionary into a model `T`, while optionally using a contextual object. Throws `UnboxError`.
+    static func performUnboxingWithDictionary<T: Unboxable>(dictionary: UnboxableDictionary, context: Any? = nil) throws -> T {
```
Could you move this so that it's together with the other `static` functions? Also, the documentation can be removed (since it's private), and remove the default value for the `context` parameter.

JohnSundell commented on Unbox.swift:
```diff
@@ -696,12 +691,64 @@ private extension Unboxer {
+
+    /// Unbox an array of JSON dictionaries into an array of `T`, while optionally using a contextual object. Throws `UnboxError`.
+    static func performUnboxingWithDictionary<T: Unboxable>(dictionaries: [UnboxableDictionary], context: Any? = nil, allowInvalidElements: Bool = false) throws -> [T] {
+        return allowInvalidElements
```
To make this line more readable (it wraps a lot in the GH interface), can you break it out into an `if` statement?

JohnSundell commented on UnboxTests.swift:
```diff
@@ -865,6 +906,134 @@ private struct UnboxTestSimpleMock: Unboxable, Equatable {
+]
+*/
+// Minified above then escaped quotes: http://www.httputility.net/json-minifier.aspx
+private let JSONMockString: String = "[{\"id\":\"56fb5031affd22691ea32b92\",\"name\":\"Clare Lancaster\",\"isActive\":true,\"age\":27,\"registered\":\"2014-07-14T07:04:50\",\"tags\":[\"magna\",\"velit\",\"qui\",\"ea\",\"fugiat\",\"est\",\"eiusmod\"],\"meta\":{\"link\":{\"self\":\"http://example.com/ex\",\"author\":\"http://example.com/exercitation\",\"collection\":\"http://example.com/enim\",\"re ...
```
Not a big fan of this, as it adds a lot of bloat to the test file. You don't need JSON, you can just create an array of dictionaries in code.

JohnSundell replied:
See `testUnboxingValueFromArray`, for an example.

JohnSundell commented:
Sorry for the delayed response :) I really like that you moved the API logic into private functions - it makes a lot of sense :+1: Some minor comments on the formatting, otherwise - great stuff! The only place where I think we need to make some changes is the test - instead of having a JSON string there I think it's a lot better to create the dictionaries in code. Let me know if you want me to help you out on any of these parts :) Thank you a lot for all the work you've done so far :rocket:

basememara commented:
Cool, I updated the code based on your comments, except for the testing. Didn't like it either, will take another swing at it..

JohnSundell commented on Unbox.swift:
```diff
@@ -655,25 +650,77 @@ private extension Unboxer {
+    }
+
+    static func performUnboxingWithDictionary<T: Unboxable>(dictionaries: [UnboxableDictionary], context: Any? = nil, allowInvalidElements: Bool = false) throws -> [T] {
+        if allowInvalidElements {
```
Awesome :+1:

JohnSundell commented:
Awesome stuff! Thanks for making those changes! :+1: Now just the testing left then we can merge this one! :smile:

basememara commented:
Having trouble getting the test to work. Here's what I'm trying:

```
    func testUnboxWithAllowInvalidElements() {
        struct Model: Unboxable {
            let successProperty: String
            let failedProperty: String
            
            init(unboxer: Unboxer) {
                self.successProperty = unboxer.unbox("success")
                self.failedProperty = unboxer.unbox("notexist")
            }
        }
        
        let array: [UnboxableDictionary] = [
            [
                "success" : "Hello"
            ],
            [
                "success" : "Unbox"
            ]
        ]
        
        guard let items: [Model] = Unbox(array, allowInvalidElements: true) else {
            XCTFail("Could not unbox collections from data")
            return
        }
        
        XCTAssert(items.count == 2, "Unbox did not return correct number of elements with invalid data")
    }
```
Commenting out the `failedProperty` makes it work, but otherwise it returns zero items. Not sure if it's the way I'm using `UnboxableDictionary` or a bug in my code? I committed it in case you want to run it.

basememara commented:
..wait think I found something... hang on..

basememara commented:
Ok there was actually a bug in my code. There were a few more places I had to add `allowInvalidElements` parameter on the dictionary side. Also, had to add it to the single-type returns (instead of just collections) so the variable can be carried all the way through. Good thing you told me to update the tests! I added a test for dictionary and another for json. Let me know if you think of anything else and I'll be glad to take another iteration. Much appreciated!

JohnSundell commented on Unbox.swift:
```diff
@@ -638,42 +633,94 @@ extension UnboxValueResolver where T: CollectionType, T: DictionaryLiteralConver
+        return try Unboxer(dictionary: dictionary, context: context, allowInvalidElements: allowInvalidElements).performUnboxingWithContext(context)
+    }
+
+    static func performUnboxingWithDictionaryAndContext<T: UnboxableWithContext>(dictionaries: [UnboxableDictionary], context: T.ContextType, allowInvalidElements: Bool = false) throws -> [T] {
```
The method name here should reflect that it takes `Dictionaries` instead of just `Dictionary`

JohnSundell commented:
Looks much better now! Although I just thought of something - why add the `allowInvalidElements` to the `Unboxer` itself? Isn't it better to keep this flag only in the code that deals with arrays, since it's only relevant for arrays? That is, in `performUnboxingWithDictionar(ies)AndContext`. That would do the job, and make a lot smaller impact on the rest of the code, right?

basememara commented:
You mean as opposed to keeping them in the `performUnboxingWithData..` functions? I needed them in the data function to parse JSON from web endpoints.

Or if you mean why keep it in the `Unboxer` class at all, I originally didn't want to, but I needed to hold that flag in memory so it was always aware of that flag. Otherwise, it exited `throwIfFailed`. If we can get that flag to be available in `throwIfFailed`, then we wouldn't need to keep it in the class.

basememara commented:
I agree tho, removing that property from `Unboxer` would bring it closer to being a struct one day and immutable. Hmm..

JohnSundell commented:
Yeah, you can put it in `performUnboxingWithData` as well, and from there call the array based one, given an array from `NSData`, since the flag only needs to be checked if one dictionary in the array failed to unbox. So the logic would be:

```
- Unbox from data (allowInvalidElements can be set).
- If the data is an array, call the dictionary API for each element (pass allowInvalidElements).
- If Unboxer did throw for one element, check allowInvalidElements before bubbling the error to the main function.
```

Makes sense?

basememara commented:
Trying to follow but not grasping yet, looking at it more..

JohnSundell commented:
@basememara Need to go to bed now (It's 00:40 here), but let's catch up tomorrow. Are you on Twitter? Maybe talking about this through IM is going to be easier, for faster turnaround! But I think we are very close to finishing this :) I'm @johnsundell if you want to catch up on Twitter.

basememara commented:
For sure, no rush. Will let it marinate and give it more thought as well. Thanks for the input again.

basememara commented:
hey! cool you were right!! I moved `allowInvalidElements` out of the `Unboxer` property and scoped it to `throwIfFailed` as a func parameter. All the tests pass and works with my local app. Can you take a closer look and make sure it jives?

JohnSundell commented on Unbox.swift:
```diff
@@ -674,34 +666,78 @@ private extension Unboxer {
     }
     
-    func performCustomUnboxingWithClosure<T>(closure: Unboxer -> T?) throws -> T {
+    func performCustomUnboxingWithClosure<T>(allowInvalidElements: Bool, closure: Unboxer -> T?) throws -> T {
```
The parameters in this function are placed in the wrong order, since `Closure` is part of the signature, it should be the first argument.

basememara replied:
I actually intentionally swapped them to allow trailing closure syntax.

JohnSundell replied:
Ah, I get it. But this is not a public API, so you don't really need that convenience. Also, the name of the method should always match the first argument. So unless you have a strong opinion about it - I think they should be switched.

basememara replied:
Right, it's just syntax sugar for internals, which we even aren't using since we're passing variables as parameters anyway. I was only thinking in case someone wanted to extend the library and has that available, but didn't feel strongly about it. Method names matching the 1st parameter is a good convention I wasn't thinking about :+1:

JohnSundell commented on Unbox.swift:
```diff
@@ -674,34 +666,78 @@ private extension Unboxer {
+    }
+
+    static func performUnboxingWithDictionaries<T: Unboxable>(dictionaries: [UnboxableDictionary], context: Any? = nil, allowInvalidElements: Bool = false) throws -> [T] {
+        return try dictionaries.map {
```
Isn't this really the only point where we need the `allowInvalidElements` flag? Do we really need to pass it along into the actual unboxing process? Shouldn't we just do a `do try catch` inside the `map` closure, and if `allowInvalidElements` is true, we just silence the error?

basememara replied:
Interesting, I'll try that.

basememara replied:
I tried it and I don't think it will work like that. Silencing the error will exclude that item from the collection completely even if it fails on one property. If it fails on one property, we still want it in the collection. I believe `allowInvalidElements` has to get to get to `throwIfFailed` in some way, but it won't this way.

JohnSundell commented:
Starts to look really good! Sorry for the slow replies, I'm traveling at the moment so have limited time to review PRs :/ Put some comments inline, let me know what you think.

basememara commented:
No worries, take your time. We're making great progress and sleeping on it definitely helps :+1:

acecilia commented:
And now we have 2 PR with the same stuff... @basememara

acecilia commented:
@basememara why don't you merge my branch in yours instead of downloading the code and committing it? This way of joining the code destroys commits history right?

basememara commented:
Yes I agree. I tried merging, creating pull requests, etc but couldn't get it done. I reverted the change just now. I believe from here you can go ahead and create a pull request against my repo, I'll accept it, then I'll place my changes on top.
